virtcontainers/kata_agent.go

```diff
@@ -429,14 +431,22 @@ func (k *kataAgent) startSandbox(sandbox *Sandbox) error {
 
 	// Start the proxy here
 	pid, uri, err := k.proxy.start(sandbox, proxyParams)
+
+	// If error occurs after kata-proxy start, rollback to kill kata-proxy process
+	defer func() {
+		if err != nil && pid > 0{
+			k.proxy.stop(sandbox, pid)
+		}
+	}()
```

Comment:
Yes, the defer is called all the time (when the function returns), and if you look further at the code in this function, there are numerous cases where we can return a non-nil error, meaning the proxy will get stopped.

Reply:
@sboeuf Since I am not when the defer function will be called, I write a simple code to test it, finally I get the this conclusion that defer function isn't called all the time after function end, for example, if a return code is called before defer in the function body, then this defer function will not be called.
But in the place, I will move defer block after the check for `err` as you suggested, because it has no effect .
If `k.proxy.start()` return error that indicates the kata-proxy process doesn't start, so we don't need to do kata-proxy rollback operation.

Reply:
Yes exactly, if the defer statement is reached, then it will be called when the function returns.
And yes, I have asked for the defer to be called after the error check of `k.proxy.start()` because we don't need any rollback if `err != nil` in this case.